source/InfoPanelState.cpp

```diff
@@ -0,0 +1,90 @@
+/* InfoPanelState.cpp
+Copyright (c) 2020 by Eric Xing
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#include "InfoPanelState.h"
+
+#include "PlayerInfo.h"
+#include "Ship.h"
+
+#include <memory>
+#include <set>
+#include <vector>
+
+using namespace std;
+
+
+
+InfoPanelState::InfoPanelState(PlayerInfo &player)
+	:ships(player.Ships()), canEdit(player.GetPlanet())
+{
+}
+
+
+int InfoPanelState::SelectedIndex() const
+{
+	return selectedIndex;
+}
+
+void InfoPanelState::SetSelectedIndex(int newSelectedIndex)
+{
+	selectedIndex = newSelectedIndex;
+}
+
+
+
+set<int> &InfoPanelState::AllSelected()
+{
+	return allSelected;
+}
```

Comment:
It might make a lot more sense to only provide `const &` access to the selection indices, and provide non-const methods e.g. `Select` & `Deselect` that take a ship and / or an index. That refactor will move maintaining this class's invariant, e.g. that when `AllSelected` is empty, SelectedIndex == -1, and otherwise SelectedIndex is a valid entry in `AllSelected`, from its current (many) place(s) in PlayerInfoPanel into this class.

Reply:
Current code is relying on `selectedIndex` and `allSelected` being independent of each other. I would prefer not to change this, as it would introduce more complexity (having to know what is going on behind the scenes in InfoPanelState).

Reply:
It feels like they should be strongly coupled, but we can defer extra refactoring until later.